### PR TITLE
Enable warnings in object libraries

### PR DESCRIPTION
**Pull request by msimberg**
Only opening now as a reminder for myself to come back to this; I have not compiled this to check how many of these there are. Inspired by https://github.com/eth-cscs/DLA-Future/pull/388#discussion_r671014072.

Additionally, we should strongly consider setting `-Wall`, with possible exceptions through `-Wno-x` but that may be a bigger task (I don't know, did someone already try it? How bad was it?).

### CODE REVIEW

**Comment by msimberg** [comment]
bors try

**Comment by rasolca** [comment]
Just as a remark: `-Wall` is already used, but is different from `-Weverything` which is anyway strongly discouraged.
https://github.com/eth-cscs/DLA-Future/blob/master/cmake/DLAF_AddTargetWarnings.cmake

**Comment by msimberg** [comment]
> Just as a remark: `-Wall` is already used, but is different from `-Weverything` which is anyway strongly discouraged.
> https://github.com/eth-cscs/DLA-Future/blob/master/cmake/DLAF_AddTargetWarnings.cmake

Very good point! I also saw that `-Wunused-local-typedefs` only really applies to typedefs (duh), and not to plain `using a::b::c`. So this would anyway not have caught all of the `using`s in #388. But it is not included in `-Wunused` nor `-Wall` so it should still be an improvement.

**Comment by rasolca** [comment]
According to gcc doc:
> `-Wunused-local-typedefs` (C, Objective-C, C++ and Objective-C++ only)
    Warn when a typedef locally defined in a function is not used. This warning is enabled by -Wall.
    
For both gcc and clang the flag is already included by `-Wall` and it covers typedefs and type aliases, but not `using namespace` and `using a::b::c` cases.

https://godbolt.org/z/G9jqhz7aP

**Comment by msimberg** [comment]
> According to gcc doc:
> 
> > `-Wunused-local-typedefs` (C, Objective-C, C++ and Objective-C++ only)
> > Warn when a typedef locally defined in a function is not used. This warning is enabled by -Wall.
> 
> For both gcc and clang the flag is already included by `-Wall` and it covers typedefs and type aliases, but not `using namespace` and `using a::b::c` cases.
> 
> https://godbolt.org/z/G9jqhz7aP

Interesting... I checked the docs first which does *not* include `-Wunused-local-typedefs` in any of `-Wall`, `-Wextra`, or `-Wunused`, but godbolt is pretty clear on that. Then the question is: why are the warnings not errors on CI? Is this the answer: https://github.com/eth-cscs/DLA-Future/blob/d9344e3390e11f2364752964a75611f2233e5a63/src/CMakeLists.txt#L141? Warnings are only added to the `DLAF` target, not the object libraries...

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
I'll try to work through these new warnings-as-errors on the side...

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
By the way, this should actually be ready to go, modulo reviews. I enabled warnings on all targets. I had to disable `-pedantic-errors` for nvcc. Otherwise I've basically only added casts or explicit integral types here and there. `std::uncaught_exception` has been deprecated in favour of `uncaught_exceptions`. That actually seems to be a C++17 thing, which GCC likely just makes available with C++14 as well... I can revert this if you'd like. Finally, I changed the variable-sized arrays `std::vector`s, except the one with bools because `std::vector<bool>` is annoying.

**Review by rasolca — APPROVED** [review]

**Comment by rasolca** [comment]
> `std::uncaught_exception` has been deprecated in favour of `uncaught_exceptions`. That actually seems to be a C++17 thing, which GCC likely just makes available with C++14 as well... I can revert this if you'd like.

@msimberg Did you already revert it? I didn't see it in the review.

File: include/dlaf/matrix/tile.h [hunk 1]
```diff
@@ -213,7 +213,7 @@ Tile<const T, device>::Tile(Tile&& rhs) noexcept
 template <class T, Device device>
 Tile<const T, device>::~Tile() {
   if (p_) {
-    if (std::uncaught_exception())
+    if (std::uncaught_exceptions() > 0)
```
Comment (msimberg) [inline include/dlaf/matrix/tile.h:216]: @rasolca I didn't revert this change yet. Should I? It may need some pragmas around it to disable the warning.

**Comment by msimberg** [comment]
bors try
